### PR TITLE
New classifier

### PR DESCRIPTION
**Pull request by RoyHirsch**
New classifier
Added early stopping (currently disabled), added logger, changed params into class, changed data_loader to be based on torchtext with bucket iterator, changed saving mode - automatically create new folder with exp num and saves log and pth files in this location.

### CODE REVIEW

File: classifier_model.py [hunk 1]
```diff
@@ -1,98 +1,71 @@
...
+        final_hidden_state = torch.cat([h_n_final_layer[i, :, :] for i in range(h_n_final_layer.shape[0])], dim=1)
+
+        attention_out = self.apply_attention(lstm_output.permute(1, 0, 2), final_hidden_state)
+        # Attention_out.shape = (batch_size, num_directions * hidden_size)
+
+        #################################### Linear #########################################
+        concatenated_vector = torch.cat([final_hidden_state, attention_out], dim=1)
+        final_feature_map = self.dropout(concatenated_vector)  # shape=(batch_size, num_directions * hidden_size)
+        final_out = self.fc(final_feature_map)
+        return self.softmax(final_out)
```
Comment (galshachaf) [inline classifier_model.py]: softmax is already used in nn.CrossEntropyLoss, no need for it here.

File: data.py [hunk 2]
```diff
@@ -27,48 +27,50 @@ def make_masks(src, tgt, device, pad=0):
...
+    build_vocab : It will first make a vocabulary or dictionary mapping all the unique words present in the train_data to an
+                  idx and then after it will use GloVe word embedding to map the index to the corresponding word embedding.
+
+    vocab.vectors : This returns a torch tensor of shape (vocab_size x embedding_dim) containing the pre-trained word embeddings.
+    BucketIterator : Defines an iterator that batches examples of similar lengths together to minimize the amount of padding needed.
+
+    """
+    # define tokenizer
+    en = English()
+    def tokenize_spacy_with_html_parsing(sentence):
```
Comment (galshachaf) [inline data.py]: does it accept a sentence, or a few sentences? I think you can't really divide it to sentences unless you strip html first.

File: train.py [hunk 3]
```diff
@@ -107,6 +110,11 @@ def save_model(self, key, save_path=None, verbose=False):
     def get_models(self):
         return self.models
 
+    def on_epoch_end(self, epoch):
```
Comment (galshachaf) [inline train.py]: you don't always want to save all the models, just the ones that improved, no?

File: train_runner.py [hunk 4]
```diff
@@ -7,71 +7,115 @@
...
+cls_opt = get_std_opt(model_cls, params)
+enc_opt = get_std_opt(model_enc, params)
+dec_opt = get_std_opt(model_dec, params)
 
-model_enc.train()
-model_dec.train()
-model_cls.train()
+early_stop = EarlyStopping(params.PATIENCE)
+for epoch in range(Params.N_EPOCHS):
+  rm.on_epoch_end(epoch)
```
Comment (galshachaf) [inline train_runner.py]: shouldn't this happen on end, not beginning?

File: train_runner.py [hunk 5]
```diff
@@ -7,71 +7,115 @@
...
-for epoch in range(N_EPOCHS):
-  run_epoch(epoch, rm, train_dataloader, model_enc, enc_opt, model_dec, dec_opt,
+  run_epoch(epoch, rm, train_iter, model_enc, enc_opt, model_dec, dec_opt,
               model_cls, cls_opt, cls_criteria, seq2seq_criteria,
-              ent_criteria, device, trans_steps=TRANS_STEPS, cls_steps=CLS_STEPS,
-              rec_lambda=REC_LAMBDA, print_interval=PRINT_INTERVAL, verbose=VERBOSE)
+              ent_criteria, params.device, params=params)
+
+  rm.on_epoch_end(epoch)
+  test_acc = evaluate(epoch, test_iter, model_enc, model_dec,
```
Comment (galshachaf) [inline train_runner.py]: I think we should save only if we improve.

File: train_runner.py [hunk 6]
```diff
@@ -7,71 +7,115 @@
...
+             model_cls, cls_criteria, seq2seq_criteria,
+             ent_criteria, params.device)
+
+  test_random_samples(test_iter, TEXT, model_enc, model_dec, model_cls, params.device,
+                      decode_func=greedy_decode_sent, num_samples=2, transfer_style=True)
+
+  # TODO - Roy - currently not in use, what metric to follow ?
```
Comment (galshachaf) [inline train_runner.py]: I think we should use accuracy

File: classifier_model.py (hunk 1 above)
Comment (RoyHirsch) [reply classifier_model.py]: True ! fixed in new commit

File: train.py (hunk 3 above)
Comment (RoyHirsch) [reply train.py]: This vanilla function saves all the three, we may extend it with more logic

File: train_runner.py (hunk 4 above)
Comment (RoyHirsch) [reply train_runner.py]: True, used of debug

File: train_runner.py (hunk 6 above)
Comment (RoyHirsch) [reply train_runner.py]: we have two measures of acc - reconstruction and classifications, with one ? an aggregation of the two ?
